Key rotation

drgora opened the pull request:
The main goal of this PR is introducing support for key rotation. Key rotation is enabled by default for `v2` sidechains' certificates, while certificates for `pre-v2` sidechains remain unchanged.

To such extent, more changes are introduced along the way:
- `mc-cryptolib` (and in turn `cctp-lib`) are updated to their most recent development versions
- `mcTestCall` is refactored (see https://github.com/HorizenOfficial/zendoo-mc-cryptolib/pull/63), and compiled as a library in `depends`
- python test utilities are updated to align with the new interface to `mcTestCall`
- few tests are updated to be more robust wrt variable performance for proofs creation and verification

drgora committed:
Zend key rotation support.

This commit adds all the bits needed to support sc key rotation.

95DDB committed:
Updated dependencies. Updated test script

95DDB committed:
Bugfix: segment_size now passed as uint32_t*

95DDB committed:
Updated dependencies

drgora committed:
Add new python test for key rotation support

drgora committed:
Refactor mcTestCall

Updates mcTestCall with the main goal of removing its duplication across different repositories. See
https://github.com/HorizenOfficial/zendoo-mc-cryptolib/pull/63 for more info.

drgora committed:
Update libzendoo reference

drgora committed:
Fix format specifiers in sidechaintypes

drgora committed:
Fix failing tests

AndreaLanfranchi commented on src/sc/proofverifier.cpp:
```diff
@@ -33,7 +33,13 @@ std::string ProofVerificationResultToString(ProofVerificationResult res)
  * @return CCertProofVerifierInput The structure containing all the data needed for the proof verification of the certificate.
  */
-CCertProofVerifierInput CScProofVerifier::CertificateToVerifierItem(const CScCertificate& certificate, const Sidechain::ScFixedParameters& scFixedParams, CNode* pfrom)
+#ifdef BITCOIN_TX
+CCertProofVerifierInput CScProofVerifier::CertificateToVerifierItem(const CScCertificate& certificate, const Sidechain::ScFixedParameters& scFixedParams, CNode* pfrom, const CCoinsViewCache* view)
+{
+    return CCertProofVerifierInput();
+}
+#else
+CCertProofVerifierInput CScProofVerifier::CertificateToVerifierItem(const CScCertificate& certificate, const Sidechain::ScFixedParameters& scFixedParams, CNode* pfrom, const CCoinsViewCache* view)
```
Please insert pre-processor directives inside function body when function signature is identical

drgora replied:
The pattern I used is quite common all over our codebase. I'd rather keep it consistent for now, and refactor all these cases in another PR if we want to.

AndreaLanfranchi replied:
Fair enough. Thank you.

AndreaLanfranchi commented on src/sc/proofverifier.cpp:
```diff
@@ -81,8 +87,15 @@ CCertProofVerifierInput CScProofVerifier::CertificateToVerifierItem(const CScCer
+            auto const& cdh = view->GetActiveCertView(certData.scId).certDataHash;
+            certData.lastCertHash = cdh.IsNull() ? CFieldElement::GetPhantomHash() : cdh;
+        }
+    }
```
Save one indentation level if possible

```suggestion
    if (view && scFixedParams.version >= 2) {
        auto const& cdh = view->GetActiveCertView(certData.scId).certDataHash;
        certData.lastCertHash = cdh.IsNull() ? CFieldElement::GetPhantomHash() : cdh;
    }
```

drgora replied:
Ok, done in 7e39749d9b01b1ee9557e5bc221d95963163f968

AndreaLanfranchi commented on src/sc/sidechaintypes.cpp:
```diff
@@ -273,10 +274,13 @@ CFieldElement CFieldElement::ComputeHash(const CFieldElement& lhs, const CFieldE
-    static CFieldElement ret{std::vector<unsigned char>(CFieldElement::ByteSize(), 0x00)};
+    // zendoo_mc_cryptolib allocates a new buffer with static lifetime on each call to zendoo_get_phantom_cert_data_hash().
+    static field_t* phantom_ptr = zendoo_get_phantom_cert_data_hash();
+    size_t size = (int)CFieldElement::ByteSize() / sizeof(unsigned char);
```
Why convert sign when operands and result are meant to be unsigned ?

AndreaLanfranchi replied:
Also not very clear to me how _staticness_ is enforced. Apparently at every call a buffer is allocated

drgora replied:
Woops, this is a leftover from a previous implementation. I updated the function entirely to avoid memcopies on each call, please see 7e39749d9b01b1ee9557e5bc221d95963163f968

AndreaLanfranchi commented on src/txmempool.cpp:
```diff
@@ -2024,23 +2024,27 @@ bool CCoinsViewMemPool::GetSidechain(const uint256& scId, CSidechain& info) cons
-                info.lastUnconfirmedReferencedHeight = -1; // -1 stands for no valid values in mempool
-            }
-            else
+        if (!mempool.mapSidechains.at(scId).mBackwardCertificates.empty()) {
```
Just a few lines above we do 
```
(mempool.mapSidechains.at(scId)
```
Maybe is worth to save a reference and avoid a second lookup on the map

drgora replied:
Yeah you're right, nice catch! I've reworked this section to remove all the unneeded accesses, thanks (7e39749d9b01b1ee9557e5bc221d95963163f968)

AndreaLanfranchi commented on src/txmempool.cpp:
```diff
@@ -2024,23 +2024,27 @@ bool CCoinsViewMemPool::GetSidechain(const uint256& scId, CSidechain& info) cons
-            else
+        if (!mempool.mapSidechains.at(scId).mBackwardCertificates.empty()) {
+
+            if (info.isNonCeasing())
```
I'd put this check above this.
```
if (!mempool.mapSidechains.at(scId).mBackwardCertificates.empty())
```
 or join conditions within a single `if`. We actualy don't care about emptyness if sc is not nonceasing

drgora replied:
Merged in 7e39749d9b01b1ee9557e5bc221d95963163f968

AndreaLanfranchi reviewed (commented):
Also ... do you guys have clang format plugins installed in your IDE/editor ?

drgora committed:
PR481 code review fixup

drgora committed:
Fixup on static init order fiasco

a-petrini commented on qa/rpc-tests/test_framework/mc_test/mc_test.py:
```diff
@@ -27,20 +27,42 @@ def __init__(self, datadir, srcdir, ps_type = "cob_marlin"):
+            if constant == True:
+                args.append("CONSTANT_PLACEHOLDER")
+            else:
+                args.append(str(constant))
```
What does happen when `constant == False`?

drgora replied:
As you suggest, when `constant == False` you actually use a constant. That is counterintuitive, and I updated `mc_test.py` to properly handle `False` as `None` in f1aa8bb237c592621a4f635666ee5a4761bbb988. Thanks!

a-petrini commented on qa/rpc-tests/test_framework/mc_test/mc_test.py:
```diff
@@ -27,20 +27,42 @@ def __init__(self, datadir, srcdir, ps_type = "cob_marlin"):
+    def _get_file_prefix(self, circ_type, constant):
+        if constant != None:
+            return str(self.ps_type) + "_" + circ_type + "_"
+        return str(self.ps_type) + "_" + circ_type + "_no_const_"
```
This could be shortened to:
```suggestion
    def _get_file_prefix(self, circ_type, constant):
        return str(self.ps_type) + "_" + circ_type + ("_" if constant != None else "_no_const_")
```

drgora replied:
Applied your suggestion to 3518be16ce52befbdfc21398187b4961582107f2

ptagl commented on qa/rpc-tests/sc_big_block.py:
```diff
@@ -140,8 +141,8 @@ def advance_sidechains_epoch(num_of_scs):
-        certVk = certMcTest.generate_params('scs', 'cert', CERT_NUM_CONSTRAINTS, SEGMENT_SIZE)
-        cswVk  = cswMcTest.generate_params('scs', 'csw', CSW_NUM_CONSTRAINTS, SEGMENT_SIZE)
+        certVk = certMcTest.generate_params('scs', 'cert', True, False, CERT_NUM_CONSTRAINTS, SEGMENT_SIZE)
+        cswVk  = cswMcTest.generate_params('scs', 'csw', True, False, CSW_NUM_CONSTRAINTS, SEGMENT_SIZE)
```
What about using parameters name to improve readability?

```suggestion
        certVk = certMcTest.generate_params('scs', 'cert', constant = True, keyrot = False, CERT_NUM_CONSTRAINTS, SEGMENT_SIZE)
        cswVk  = cswMcTest.generate_params('scs', 'csw', constant = True, keyrot = False, CSW_NUM_CONSTRAINTS, SEGMENT_SIZE)
```

drgora replied:
Good point, done with f663cac6ada8f2b0e099497b23ac96371e611b14

ptagl commented on qa/rpc-tests/sc_cert_base.py:
```diff
@@ -148,15 +148,15 @@ def run_test(self):
         quality = 10
         proof = mcTest.create_test_proof(
-            "sc1", scid_swapped, epoch_number, quality, MBTR_SC_FEE, FT_SC_FEE, epoch_cum_tree_hash, constant, [addr_node1], [bwt_amount])
+            "sc1", scid_swapped, epoch_number, quality, MBTR_SC_FEE, FT_SC_FEE, epoch_cum_tree_hash, None, constant, [addr_node1], [bwt_amount])
```
Can we use arguments name to avoid using `None` when calling `create_test_proof()` as in other tests?

That should apply also to other calls to the same function in this file.

```suggestion
            "sc1", scid_swapped, epoch_number, quality, MBTR_SC_FEE, FT_SC_FEE, epoch_cum_tree_hash, constant = constant, pks = [addr_node1], amounts = [bwt_amount])
```

drgora replied:
@a-petrini did this in 64a018ce829d1a900692a9e41d2589cbb3518939, thanks a lot!

a-petrini commented on qa/rpc-tests/test_framework/blockchainhelper.py:
```diff
@@ -173,7 +173,7 @@ def create_sidechain(self, sc_name, sidechain_version, optional_parameters = {},
 
-        sc_input = SidechainCreationInput.from_rpc_args(self, sc_name, optional_parameters)
+        sc_input = SidechainCreationInput.from_rpc_args(self, sc_name, sidechain_version, optional_parameters)
         sc_input.version = sidechain_version
```
This should no longer be needed, as `from_rpc_args` passes the value `sidechain_version` to `SidechainCreationInput` constructor. All in all, maybe it was possible to correctly set the SC version by passing a proper key-value in the `optional_parameters` dictionary

drgora replied:
I removed line 177 in f1aa8bb237c592621a4f635666ee5a4761bbb988. We may update to only use `optional_parameters` in future.

ptagl commented on qa/rpc-tests/sc_csw_memcleanup_split.py:
```diff
@@ -152,15 +152,15 @@ def run_test(self):
 
         proof = certMcTest.create_test_proof(
-            "sc1", scid_swapped, epoch_number, quality, MBTR_SC_FEE, FT_SC_FEE, epoch_cum_tree_hash, constant, [addr_node1], [bt_amount])
+            "sc1", scid_swapped, epoch_number, quality, MBTR_SC_FEE, FT_SC_FEE, epoch_cum_tree_hash, None, constant, [addr_node1], [bt_amount])
```
```suggestion
            "sc1", scid_swapped, epoch_number, quality, MBTR_SC_FEE, FT_SC_FEE, epoch_cum_tree_hash, constant = constant, pks = [addr_node1], amounts = [bt_amount])
```

drgora replied:
@a-petrini did this in 64a018ce829d1a900692a9e41d2589cbb3518939, thanks a lot!

ptagl commented on qa/rpc-tests/sc_getcertmaturityinfo.py:
```diff
@@ -77,7 +77,7 @@ def run_test(self):
         cert1_sc1: str = test_helper.send_certificate(v1_sc1_name, 10) # Epoch 0
         cert1_sc2: str = test_helper.send_certificate(v2_ceasing_sc2_name, 10) # Epoch 0
-        cert1_sc3: str = test_helper.send_certificate(v2_non_ceasing_sc3_name, 0) # Epoch 0
+        cert1_sc3: str = test_helper.send_certificate(v2_non_ceasing_sc3_name, 0, self.nodes[0].getblockcount()-1) # Epoch 0, reference one block in the past
```
Why such a change was needed? Did the test stop working for any reason?

a-petrini commented on qa/rpc-tests/test_framework/blockchainhelper.py:
```diff
@@ -304,9 +298,4 @@ def send_certificate(self, sc_name, quality, referenced_height = None):
-
-        if is_non_ceasing_sidechain:
-            sc_info["last_referenced_height"] = ref_height
-
```
Is there a reason for no longer updating there values in the `sc_info` dictionary after submitting the certificate?

drgora replied:
My mistake, thanks for catching this. Restored in 3518be16ce52befbdfc21398187b4961582107f2

a-petrini commented on qa/rpc-tests/sc_cert_getblocktemplate.py:
```diff
@@ -185,16 +186,15 @@ def run_test_with_scversion(self, scversion, ceasable = True):
-                sc_name, scid_swapped, epoch_number, quality, MBTR_SC_FEE, FT_SC_FEE, epoch_cum_tree_hash, constant,
-                [addr_node1], [bwt_amount])
+                sc_name, scid_swapped, epoch_number, quality, MBTR_SC_FEE, FT_SC_FEE, epoch_cum_tree_hash,
+                prev_cert_hash = prev_cert_data_hash if scversion >= 2 else None, constant = constant, pks = [addr_node1], amounts = [bwt_amount])
```
Thanks for uniforming the proof generation among SC versions. However, now the only difference is in the increase of quality for ceasable SC, so everything else can be removed from the `if..else` blocks

drgora replied:
Nice catch! Done in f1aa8bb237c592621a4f635666ee5a4761bbb988, thanks.

ptagl commented on qa/rpc-tests/test_framework/blockchainhelper.py:
```diff
@@ -89,7 +89,7 @@ def __init__(self, blockchainHelper, name):
         return self.version == 2 and self.withdrawalEpochLength == 0
 
-    def from_rpc_args(blockchainHelper, name, args):
+    def from_rpc_args(blockchainHelper, name, version, args):
```
Why do we need an additional `version` parameter? Can't we rely on the one already available in the `args` structure?

ptagl replied:
Already addressed with [this comment](https://github.com/HorizenOfficial/zen/pull/481#discussion_r999213311).

ptagl commented on qa/rpc-tests/test_framework/blockchainhelper.py:
```diff
@@ -173,7 +173,7 @@ def create_sidechain(self, sc_name, sidechain_version, optional_parameters = {},
 
-        sc_input = SidechainCreationInput.from_rpc_args(self, sc_name, optional_parameters)
+        sc_input = SidechainCreationInput.from_rpc_args(self, sc_name, sidechain_version, optional_parameters)
         sc_input.version = sidechain_version
```
Considering the new behavior of the `SidechainCreationInput` contructor, can't we remove this line?

ptagl replied:
Already addressed with [this comment](https://github.com/HorizenOfficial/zen/pull/481#discussion_r999213311).

a-petrini commented on qa/rpc-tests/walletbackup.py:
```diff
@@ -238,8 +238,21 @@ def run_test_with_scversion(self, scversion, ceasable = True):
+                                                 MBTR_SC_FEE,
+                                                 FT_SC_FEE,
+                                                 epoch_cum_tree_hash,
+                                                 prev_cert_hash,
```
Is it possible to do as in [qa/rpc-tests/sc_cert_getblocktemplate.py]? This would also remove the if at lines 242-243
```suggestion
                                                 prev_cert_hash if scversion >=2 else None,
```

drgora replied:
Done in f1aa8bb237c592621a4f635666ee5a4761bbb988

ptagl commented on qa/rpc-tests/test_framework/util.py:
```diff
@@ -611,12 +620,14 @@ def advance_epoch(mcTest, node, sync_call,
 
     scid_swapped = str(swap_bytes(scid))
 
+    print(constant)
```
I guess this print was used only for debugging purpose, can we remove it now?

drgora replied:
Right, this should have been removed, nice catch! f663cac6ada8f2b0e099497b23ac96371e611b14

ptagl commented on src/txmempool.cpp:
```diff
@@ -2013,23 +2013,28 @@ bool CCoinsViewMemPool::GetSidechain(const uint256& scId, CSidechain& info) cons
+        if (info.isNonCeasing() && !sc.mBackwardCertificates.empty()) {
+            const uint256& topQualHash = sc.GetTopQualityCert()->second;
             const CScCertificate& certTopQual = mempool.mapCertificate[topQualHash].GetCertificate();
+            info.lastTopQualityCertView.certDataHash = certTopQual.GetDataHash(info.fixedParams);
+            info.lastTopQualityCertView.forwardTransferScFee = certTopQual.forwardTransferScFee;
+            info.lastTopQualityCertView.mainchainBackwardTransferRequestScFee = certTopQual.mainchainBackwardTransferRequestScFee;
```
It's quite clear to me why we need to include the `certDataHash` into the sidechain info, but I don't understand why we do the same for `forwardTransferScFee` or `mainchainBackwardTransferRequestScFee`.

As far as I remember we only take into account the fees published with the last block and not the fees that are set by unconfirmed certificates.

ptagl commented on src/sc/sidechaintypes.cpp:
```diff
@@ -273,11 +278,18 @@ CFieldElement CFieldElement::ComputeHash(const CFieldElement& lhs, const CFieldE
+    phantom_element.SetByteArray(buf);
+    return phantom_element;
 }
 #endif
```
I think that the new behavior of this function is more correct than the previous one, but I wonder if it breaks the compatibility with previous versions.

We should perform additional checks to make sure that it's safe to do so.

drgora replied:
You're right,this indeed breaks compatibility. f663cac6ada8f2b0e099497b23ac96371e611b14 restores this zero hash for compatibility, and also implements `GetPhantomHash` properly to return the phantom hash as expected.

a-petrini commented on qa/rpc-tests/sc_cert_nonceasing.py:
```diff
@@ -46,13 +46,13 @@ def setup_network(self, split=False):
-        proof = self.mcTest.create_test_proof(PARAMS_NAME, scid_swapped, epoch_number, quality, mbtr_fee, ft_fee, epoch_cum_tree_hash, self.constant, [bt["address"]], [bt["amount"]])
+        ep, epoch_cum_tree_hash, prev_cert_hash = get_epoch_data(scid, self.nodes[node_idx], 0, True, ref_height)
+        proof = self.mcTest.create_test_proof(PARAMS_NAME, scid_swapped, epoch_number, quality, mbtr_fee, ft_fee, epoch_cum_tree_hash, prev_cert_hash,
+            constant = self.constant, pks = [bt["address"]], amounts = [bt["amount"]])
```
```suggestion
        proof = self.mcTest.create_test_proof(PARAMS_NAME, scid_swapped, epoch_number, quality, mbtr_fee, ft_fee, epoch_cum_tree_hash, prev_cert_hash = prev_cert_hash,
            constant = self.constant, pks = [bt["address"]], amounts = [bt["amount"]])
```

ptagl commented on src/gtest/tx_creation_utils.cpp:
```diff
@@ -753,7 +753,7 @@ void BlockchainTestManager::GenerateSidechainTestParameters(ProvingSystem provin
     CctpErrorCode errorCode;
     zendoo_generate_mc_test_params(
-        circuitType, provingSystem, 1 << 10,
+        circuitType, provingSystem, 1 << 10, false, //FIXME?
```
Maybe we can expose the argument `with_prev_hash` (currently set to false) to the callers of `GenerateSidechainTestParameters()` as optional.

That would also let us to extend the unit tests currently available in `src/gtest/test_libzendoo.cpp`.

drgora replied:
Good point, I implemented your suggestion in f663cac6ada8f2b0e099497b23ac96371e611b14, which also adds key rotation specific tests in `test_libzendoo.cpp`

ptagl commented on qa/rpc-tests/sc_keyrot.py:
```diff
@@ -0,0 +1,226 @@
+
+    def run_test(self):
+        '''
+        '''
```
Can we add a summary description of what is tested here?

drgora replied:
Woops, right, I forgot this, thanks. Added in f663cac6ada8f2b0e099497b23ac96371e611b14

drgora committed:
Minor improvements to Python tests

ptagl commented on qa/rpc-tests/sc_keyrot.py:
```diff
@@ -0,0 +1,226 @@
+                                  0, # 0 means that PHANTOM_HASH will be used
+                                  False
+                                 )
+
```
I see here that we test passing the PHANTOM_HASH for the proof of the first certificate. Should we add more tests to cover also the generation of proof for certificates with epoch other than the first one (so that we don't pass the PHANTOM_HASH) or this scenario is already taken into account in other Python tests?

drgora replied:
Such scenario is indeed tested in all the other tests that send v2 sc certificates, such as `sc_cert_nonceasing.py` and `sc_getcertmaturityinfo.py`.

a-petrini committed:
Refactoring of all create_test_proof() calls in python tests

We updated all the create_test_proof() in Python test to explicitly assign non-positional
arguments to their respective parameter names. Also, each function parameter is now
written on a separate line to improve readability.

a-petrini committed:
fix test broken in refactoring

drgora committed:
Address more PR481 feedback

Specifically, this commit restores zero-hash default value for pre-SC fork scCumTreeHash.
Also, it adds more gtests for key rotation proof creation and verficiation.
Finally, it adds documentation and implements minor cosmetic changes to Python tests.

drgora committed:
Validate scfees only against certificates in the blockchain.

This commit also adds a test section (sc_cert_nonceasing.py) that verifies the expected behavior with v2 non ceasing
sidechains.
Finally, some superseded fields are removed from CSidechain.

drgora committed:
Panic if block referenced by certificate in mempool cannot be found.
